[OSDOCS-3535] Release note addition for pod security admission (#47151)
openshift/openshift-docs: release_notes/ocp-4-11-release-notes.adoc

For the full list of OIDC providers, see xref:../authentication/identity_providers/configuring-oidc-identity-provider.adoc#identity-provider-oidc-supported_configuring-oidc-identity-provider[Supported OIDC providers].

[id="ocp-4-11-auth-pod-security-admission"]
==== Pod security admission

link:https://kubernetes.io/docs/concepts/security/pod-security-admission/[Pod security admission] is now enabled on {product-title}. 

Pod admission is enforced by both pod security and security context constraint (SCC) admissions. Pod security admission runs globally with "privileged" enforcement and "restricted" audit logging and API warnings.

> A controller monitors the xref:../authentication/managing-security-context-constraints.adoc[SCC-related permissions] of service accounts in user-created namespaces and automatically labels these namespaces with link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security admission] `warn` and `audit` labels.

Reviewer: We need to also clarify that the PSA does not work on the namespace prefixed with openshift 
Also, add the info in the release notes that the PSA provides a feature to be enabled and disabled. 
Therefore the list of namespaces that the PSA cannot work at all. 
Also, we ought to make clear that now the "openshift" namespaces are Previleged but we plan to enforce them to "Restricted" in the future so that OCP users can plan better their changes and start to avoid using them to install solutions. 

Therefore, I'd like to suggest we add **something like:** 

```suggestion
A controller monitors the xref:../authentication/managing-security-context-constraints.adoc[SCC-related permissions] of service accounts in user-created namespaces and automatically labels these namespaces with link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security admission] `warn` and `audit` labels.

The SCC synchronization is disabled by default for user-created namespaces that are named with the prefix openshift-. Namespaces that have this prefix are typically system namespaces; by convention, a controller exists to manage them.

This new component also introduces a feature which allows you to enable SCC synchronization in namespaces that have the prefix openshift- by setting the value of the `security.openshift.io/scc.podSecurityLabelSync` label to true. 

However, some namespaces are restricted and do not allow SCC synchronization. These namespaces are system namespaces. You can verify the [list](https://github.com/openshift/cluster-policy-controller/blob/release-4.11/pkg/psalabelsyncer/podsecurity_label_sync_controller.go#L42-L111) of namespaces those is not allowed the synchronization. In this release, these ones are by default enforced with "Privileged".  However, it is planned for the next release to change the performance of these to "Restricted". 

--

```

c/c @stlaz

Reviewer: The namespaces prefixed with "openshift" are NOT system one (a least up 4.10). Then, if we want to define it as a rule now it is a new change defined and another point that must be highlighted in the release notes. See that so far we have users and also OLM catalog using these to install Operators. 

Example "openshift-operators" is a default namespace provided in OCP to install operators with the OLM console.

Reviewer 2: I'm going to assume that by PSa you mean the label syncing. 

> the PSA does not work on the namespace prefixed with openshift by default

Keep the release note brief, point to the docs, where we're already documenting this, I believe

> Also, add the info in the release notes that the PSA provides a feature to be enabled and disabled.

Indeed, we might want to have a sentence that this works on an opt-in/opt-out basis that is decribed further in the docs

> make clear that for 4.11 the prefixed "openshift" namespaces have no enforcement for restricted but it is planned to make the restricted on the future release

Good point, if it's not here yet, we should state that our intention for an upcoming release is to be enforcing the "restricted" profile by default and having the automatic labelsyncing work with enforcement labels.

To improve workload security in accordance with the "restricted" pod security profile, this release introduces SCCs that enforce pod security in accordance with new pod security admission controls. These SCCs are:

* `restricted-v2`
* `hostnetwork-v2`
* `nonroot-v2`

They correspond to older, similarly named SCCs, but with the following enhancements:

* `ALL` capabilities are dropped from containers.
* The `NET_BIND_SERVICE` capability can be added explicitly.
* `seccompProfile` is set to `runtime/default` by default.

> * `allowPrivilegeEscalation` must be unset or set to `false` in security contexts. 

Reviewer: @stlaz `allowPrivilegeEscalation = false` is valid for all new policies ?

> * `allowPrivilegeEscalation` must be unset or set to `false` in security contexts. 

Reviewer: IMO we need a more detailed DELTA between v1 and v2. On top of that. we might want to be more specific here. See:

> The `NET_BIND_SERVICE` capability can be added explicitly. 

For restricted-v2 we cannot add this one and the text give the impression that it is also valid for restricted-v2.

Suggestions:
a) Ensure that we clarify that:

- V2 does not permit allowPrivilegeEscalation=true (empty or false is compatible with v1 SCC and therefore works on previous OCP versions < 4.11)
- V2 requires you to either leave the dropped capabilities empty, or set it to ALL, or either allow a subset. Note that by being accepted as restricted-v2 the SCC will drop ALL always. V1 only dropped KILL,MKNOD,SETUID,SETGID capabilities while V2 drops ALL
- V2 requires you to either leave seccompProfile empty or set it to runtime/default (empty is compatible with v1 and works on previous OCP versions < 4.11 )

(ihmo) would be great if we could create a table that compares each policy from v1 to v2

Reviewer: c/c @stlaz

> * `ALL` capabilities are dropped from containers.
> * The `NET_BIND_SERVICE` capability can be added explicitly.
> * `seccompProfile` is set to `runtime/default` by default.
> * `allowPrivilegeEscalation` must be unset or set to `false` in security contexts. 

Reviewer: IMO we need a more detailed DELTA between v1 and v2. On top of that. we might want to be more specific here it allows those that develop solutions for OCP and cluster admins have better info to check if/when/how they are or are not impacted. See:

**Suggestions:**

Ensure that we clarify that: 

- a) we need explicitly to let the users know what was dropped before and now is not. That is important for those that need to evaluate/troubleshoot if a solution running on the cluster might or not be impacted. (i.e the workload was using capabilities that were NOT dropped before but will not which will not allow the workload to do what it must. )) 

- b) Make clear the requirements to be qualified as v2 which were not before with v1 for each case. 

```suggestion
* `ALL` capabilities are dropped from containers. Note that for the previous SCC policies for nonroot, restricted and hostnetwork ONLY `KILL`,`MKNOD`,`SETUID`,`SETGID` capabilities are dropped while the (v2) policies drops ALL. _(The new policies require you to either leave the dropped capabilities empty, or set it to ALL, or either allow a subset explicitly. )_
* The `NET_BIND_SERVICE` capability can be added explicitly.
* `seccompProfile` is set to `runtime/default` by default. V2 SCC policies require you to either leave seccompProfile empty or set it to runtime/default. _(Note that for the previous SCC policies you MUST leave empty.)_
* `allowPrivilegeEscalation` must be unset or set to `false`. 
```

c/c @stlaz

Author: @stlaz Any thoughts about this?

Reviewer 2: Let's stay as brief as possible. I think a reasonable compromise is:

```suggestion
* `ALL` capabilities are dropped from containers. Previously only a small set of capabilities would have been dropped.
* The `NET_BIND_SERVICE` capability can be added explicitly.
* `seccompProfile` gets defaulted to `runtime/default` if unset. In previous releases, this field was required to stay empty.
* `allowPrivilegeEscalation` must be unset or set to `false` in security contexts. Previously it was also allowed to be set to `true`.
```

Reviewer 2: @maxwelldb as per discussion with @camilamacedo86, let's include the list of previously dropped caps:
```
`ALL` capabilities are dropped from containers. Previously only `KILL`, `MKNOD`, `SETUID`, `SETGID` were being dropped.
```

Author: ```suggestion
* `ALL` capabilities are dropped from containers. Previously, only a small set of capabilities were dropped.
* The `NET_BIND_SERVICE` capability can now be added explicitly.
* `seccompProfile` is defaulted to `runtime/default` if unset. In previous releases, this field had to be empty.
* `allowPrivilegeEscalation` must be unset or set to `false` in security contexts. Previously, a `true` value was allowed.
```
@stlaz

New clusters allow use of the `restricted-v2` SCC for any authenticated user in place of the `restricted` SCC; the `restricted` SCC is no longer available to users of new clusters, unless the access is explicitly granted. Clusters that are upgraded to 4.11 from previous versions allow both `restricted-v2` and `restricted` SCCs. 

> 

Author: ```suggestion

You can enable synchronization for most namespaces and disable synchronization for all namespaces. 

For this release, namespaces that are prefixed with `openshift-` will not have restricted enforcement. Restricted enforcement for such namespaces is planned for inclusion in a future release.

```

For the full list of OIDC providers, see xref:../authentication/identity_providers/configuring-oidc-identity-provider.adoc#identity-provider-oidc-supported_configuring-oidc-identity-provider[Supported OIDC providers].

[id="ocp-4-11-auth-pod-security-admission"]

> ==== Pod security admission autolabeling

Reviewer 2: just "Pod Security Admission"

Author: ```suggestion
==== Pod security admission
```

> link:https://kubernetes.io/docs/concepts/security/pod-security-admission/[Pod security admission] is now enabled on {product-title}. 

Author: ```suggestion
link:https://kubernetes.io/docs/concepts/security/pod-security-admission/[Pod security admission] is now enabled on {product-title}. 

Clusters support pod security admission, as well as SCC admission, for pod admission. Pod security admission runs globally with "privileged" enforcement and "restricted" audit logging and API warnings.
```

> If the pod security admission plug-in is activated, {product-title} preserves pod workloads despite changes to the pod security API. A cluster performs this action by monitoring administrator-labeled namespaces for changes and mapping link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security standards] to xref:../authentication/managing-security-context-constraints.adoc[security context constraints].

Reviewer 2: `If the pod security admission plug-in is activated, {product-title} preserves pod workloads despite changes to the pod security API`
I don't understand this, and I think we can simply leave it out, because pod security admission is always on.

Author: ```suggestion
Clusters now monitor administrator-labeled namespaces for changes and map link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security standards] to xref:../authentication/managing-security-context-constraints.adoc[security context constraints].
```

> A controller monitors service account xref:../authentication/managing-security-context-constraints.adoc[SCC-related permissions] in user-created namespaces and automatically labels these namespaces with link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security admission] `warn` and `audit` labels.

Author: @stlaz As in:

* The permissions of a service account as a concept
* The permissions of all service accounts as concrete objects

At the moment, it works grammatically.

> A controller monitors service account xref:../authentication/managing-security-context-constraints.adoc[SCC-related permissions] in user-created namespaces and automatically labels these namespaces with link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security admission] `warn` and `audit` labels.

Author: ```suggestion
A controller monitors the xref:../authentication/managing-security-context-constraints.adoc[SCC-related permissions] of service accounts in user-created namespaces and automatically labels these namespaces with link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security admission] `warn` and `audit` labels.
```

> You can learn more about admission controller configuration in the link:https://kubernetes.io/docs/tasks/configure-pod-container/enforce-standards-admission-controller/#configure-the-admission-controller[Kubernetes documentation].

Reviewer 2: I don't think we need this here.

Author: ```suggestion
```

> The `privileged` pod security profile is enforced, and the `restricted` profile is used for warnings and audits.

Reviewer 2: `OpenShift 4.11 is using the Pod Security admission[link] for pod admission, alongside with the SCC admission. Pod Security admission runs globally with "privileged" enforcement and "restricted" audit logging and API warnings`.

I would expect this release note to start something like this.

> This release also introduces new security context constraints (SCCs) that enforce pod security in accordance with new pod security admission controls. These SCCs are:

Reviewer 2: I think we would like to mention that this change was to improve the workloads security to be able to reach the "restricted" pod security profile, something like:

`In order to improve workloads security according to the "restricted" pod security profile, this release also introduces..`

Author: ```suggestion
To improve workload security by using the "restricted" pod security profile, this release introduces security context constraints (SCCs) that enforce pod security in accordance with new pod security admission controls. These SCCs are:
```

* `restricted-v2`
* `hostnetwork-v2`
* `nonroot-v2`

They correspond to older, similarly named SCCs, but with the following enhancements:

* `ALL` capabilities are dropped from containers.
* The `NET_BIND_SERVICE` capability can be added explicitly.
* `seccompProfile` is set to `runtime/default` by default.
* `allowPrivilegeEscalation` must be unset or set to `false` in security contexts. 

New clusters allow use of the `restricted-v2` SCC for any authenticated user in place of the `restricted` SCC; the `restricted` SCC is no longer available to users of new clusters, unless the access is explicitly granted. Clusters that are upgraded to 4.11 from previous versions allow both `restricted-v2` and `restricted` SCCs. 

> // TODO Add link

Author: Yep. #47647 has to be merged before a link can be added here.

> // TODO Add link

Author: ```suggestion
// TODO: Add anchor
// For more information, see xref:../authentication/managing-and-understanding-pod-security-admission.adoc[Understanding and managing pod security admission].
```

> New clusters use the `restricted-v2` SCC in place of the `restricted` SCC that is available to any authenticated user. Clusters that are upgraded to 4.11 from previous versions allow both `restricted-v2` and `restricted` SCCs. 

Reviewer 2: > New clusters use the `restricted-v2` SCC in place of the `restricted` SCC that is available to any authenticated user.

not being native, does this unambiguously mean that the `restricted` is no longer available to authenticated users in freshly installed clusters?

Author: ```suggestion
New clusters use the `restricted-v2` SCC in place of the `restricted` SCC; the `restricted` SCC is no longer available to users of new clusters. Clusters that are upgraded to 4.11 from previous versions allow both `restricted-v2` and `restricted` SCCs. 
```

For the full list of OIDC providers, see xref:../authentication/identity_providers/configuring-oidc-identity-provider.adoc#identity-provider-oidc-supported_configuring-oidc-identity-provider[Supported OIDC providers].

[id="ocp-4-11-auth-pod-security-admission"]
==== Pod security admission

link:https://kubernetes.io/docs/concepts/security/pod-security-admission/[Pod security admission] is now enabled on {product-title}. 

> Clusters support pod security admission, as well as SCC admission, for pod admission. Pod security admission runs globally with "privileged" enforcement and "restricted" audit logging and API warnings.

Reviewer 2: ```suggestion
Pod admission is therefore enforced by both the Pod Security and the SCC admissions. Pod security admission runs globally with "privileged" enforcement and "restricted" audit logging and API warnings.
```
- to maybe go a bit easy on the repetition of "admission" here

Author: ```suggestion
Pod admission is enforced by both pod security and the SCC admissions. Pod security admission runs globally with "privileged" enforcement and "restricted" audit logging and API warnings.
```

Reviewer 2: again, not a native speaker, but I assume that you'd have to use `the` with both the subjects, or neither - `pod security and SCC admissions`

Author: Fixed!

> Clusters now monitor administrator-labeled namespaces for changes and map link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security standards] to xref:../authentication/managing-security-context-constraints.adoc[security context constraints].

Reviewer 2: `A controller monitors Service Account SCC-related permissions in user-created namespaces and automatically labels these namespaces with PSa `warn` and `audit` labels` (while keeping all the links :slightly_smiling_face:  )

Author: ```suggestion
A controller monitors service account xref:../authentication/managing-security-context-constraints.adoc[SCC-related permissions] in user-created namespaces and automatically labels these namespaces with link:https://kubernetes.io/docs/concepts/security/pod-security-standards/[pod security admission] `warn` and `audit` labels.
```

> The `privileged` pod security profile is enforced, and the `restricted` profile is used for warnings and audits.

Reviewer 2: Remove this, this is already mentioned here:
`Clusters support pod security admission, as well as SCC admission, for pod admission. Pod security admission runs globally with "privileged" enforcement and "restricted" audit logging and API warnings.`

Author: ```suggestion
```

> To improve workload security by using the "restricted" pod security profile, this release introduces security context constraints (SCCs) that enforce pod security in accordance with new pod security admission controls. These SCCs are:

Reviewer 2: There is no "using" the "restricted profile. The new SCC really only allow to reach the restrictivity of the profile.

Author: ```suggestion
To improve workload security in accordance with the "restricted" pod security profile, this release introduces security context constraints (SCCs) that enforce pod security in accordance with new pod security admission controls. These SCCs are:
```

* `restricted-v2`
* `hostnetwork-v2`
* `nonroot-v2`

They correspond to older, similarly named SCCs, but with the following enhancements:

* `ALL` capabilities are dropped from containers.
* The `NET_BIND_SERVICE` capability can be added explicitly.
* `seccompProfile` is set to `runtime/default` by default.
* `allowPrivilegeEscalation` must be unset or set to `false` in security contexts. 

> New clusters use the `restricted-v2` SCC in place of the `restricted` SCC; the `restricted` SCC is no longer available to users of new clusters. Clusters that are upgraded to 4.11 from previous versions allow both `restricted-v2` and `restricted` SCCs. 

Reviewer 2: ```suggestion
New clusters allow use of the `restricted-v2` SCC for any authenticated user in place of the `restricted` SCC; the `restricted` SCC is no longer available to users of new clusters, unless the access is explicitly granted. Clusters that are upgraded to 4.11 from previous versions allow both `restricted-v2` and `restricted` SCCs. 
```
